### PR TITLE
add module concept

### PR DESCRIPTION
**Pull request by sns-seb**
this PR introduces the concept of module as a group of components
this will allow reducing length of the ComponentLevel classes and in the near future implement lazy start of components and/or description of implicit component dependencies
this PR also includes a few module to demo usage and as a proof-of-concept

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/batch/BatchWsModule.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.batch;
+
+import org.sonar.core.component.Module;
+import org.sonar.server.computation.ws.SubmitReportAction;
+
+public class BatchWsModule extends Module {
+  @Override
+  protected void configureModule() {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/batch/BatchWsModule.java]: :new: MAJORProtected method 'configureModule(...)' is never used.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUnusedProtectedMethod)

File: server/sonar-server/src/main/java/org/sonar/server/search/EsSearchModule.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.search;
+
+import org.sonar.core.component.Module;
+import org.sonar.server.es.EsClient;
+
+public class EsSearchModule extends Module {
+  @Override
+  protected void configureModule() {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/search/EsSearchModule.java]: :new: MAJORProtected method 'configureModule(...)' is never used.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUnusedProtectedMethod)

File: server/sonar-server/src/main/java/org/sonar/server/platform/platformlevel/PlatformLevel4.java [hunk 3]
```diff
@@ -306,7 +299,7 @@ public PlatformLevel4(PlatformLevel parent, List<Object> level4AddedComponents)
   }
 
   @Override
-  public PlatformLevel configure() {
+  protected void configureLevel() {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/platform/platformlevel/PlatformLevel4.java]: :new: MAJORThis method has 387 lines, which is greater than the 100 lines authorized. Split it into smaller methods.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS138)

File: server/sonar-server/src/main/java/org/sonar/server/db/migrations/MigrationStepModule.java [hunk 4]
```diff
@@ -73,78 +73,77 @@
 import org.sonar.server.db.migrations.v52.FeedProjectLinksComponentUuid;
 import org.sonar.server.db.migrations.v52.MoveProjectProfileAssociation;
 
-import java.util.List;
+public class MigrationStepModule extends Module {
+  @Override
+  protected void configureModule() {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/db/migrations/MigrationStepModule.java]: :new: MAJORProtected method 'configureModule(...)' is never used.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUnusedProtectedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 new issues:
* 4 major

Details:
* [Protected method 'configureModule(...)' is never used.](https://github.com/SonarSource/sonarqube/blob/283df66342aa71314a090323fe65a59927fc0b05/server/sonar-server/src/main/java/org/sonar/server/batch/BatchWsModule.java#L27)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUnusedProtectedMethod)
* [Protected method 'configureModule(...)' is never used.](https://github.com/SonarSource/sonarqube/blob/283df66342aa71314a090323fe65a59927fc0b05/server/sonar-server/src/main/java/org/sonar/server/db/migrations/MigrationStepModule.java#L78)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUnusedProtectedMethod)
* [This method has 387 lines, which is greater than the 100 lines authorized. Split it into smaller methods.](https://github.com/SonarSource/sonarqube/blob/283df66342aa71314a090323fe65a59927fc0b05/server/sonar-server/src/main/java/org/sonar/server/platform/platformlevel/PlatformLevel4.java#L302)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS138)
* [Protected method 'configureModule(...)' is never used.](https://github.com/SonarSource/sonarqube/blob/283df66342aa71314a090323fe65a59927fc0b05/server/sonar-server/src/main/java/org/sonar/server/search/EsSearchModule.java#L27)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUnusedProtectedMethod)

**Comment by simonbrandhof** [comment]
There's a conflict in terminology. SQ also deals with "components" and "modules" when inspecting source code. 
Unfortunately I don't have any proposal.

**Comment by sns-seb** [comment]
ComponentModule?

**Comment by simonbrandhof** [comment]
You're talking about which module of components ? :-(

**Comment by julienlancelot** [comment]
What about using 'aggregator' ?

**Comment by simonbrandhof** [comment]
Let's merge and improve naming later.

**Comment by sns-seb** [comment]
I don't think there is any way around it:
* "component" is a very generic name and is used in tones of contexts, so does "module"
* here, we are in the IOC context, both terms are quite common (see Dagger, Guice for "module"), I don't see how we could benefit from inventing words here

Overall and from a practical point of view, I don't think this is a problem anyway.
"Component" and "Module" will be used as part of class or interface names, I don't see how it could be confused with a component or a module as an SQ source inspection unit (why would an interface or class be used to represent the latter). Regarding other usage of "component" or "module" (in code, methods names, ...), those will occur only in precisely located places (ie. in the code and package that handles IOC)

File: sonar-core/src/main/java/org/sonar/core/component/Module.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    if (object != null) {
+      container.addComponent(object, singleton);
+    }
+  }
+
+  protected <T> T getComponentByType(Class<T> tClass) {
+    return container.getComponentByType(tClass);
+  }
+
+  protected void add(Object... objects) {
```
Comment (simonbrandhof) [inline sonar-core/src/main/java/org/sonar/core/component/Module.java]: Sounds misleading to have both add(Object, boolean) and add(Object...) that implicitly sets the boolean to true. 
Moreover don't you think that these protected methods aim to duplicate ComponentContainer ? An alternative could be: ```protected abstract void configureModule(ComponentContainer container)```

Comment (sns-seb) [inline sonar-core/src/main/java/org/sonar/core/component/Module.java]: `add(Object, boolean)` is currently never used. It created it only to provide support for the current use of `ComponentContainer.add(Object, boolean)` in `PlatformLevel`. Btw, the `ComponentContainer` method is used only once with a false parameter and I don't think it is even appropriate https://github.com/SonarSource/sonarqube/blob/master/server/sonar-server/src/main/java/org/sonar/server/platform/platformlevel/PlatformLevel4.java#L424.

I'm not very happy with the idea of `protected abstract void configureModule(ComponentContainer container)`.

First, it creates a non necessary coupling between subclasses of `Module` and the class `ComponentContainer`. Implementation only cares to add classes or objects to the module. How this objects are in turn added to the container is implementation specific. Side-benefit: if implementation is someday not based on `ComponentContainer` but, eg. on PicoContainer directly, Module subclasses will not change.

Second, `ComponentContainer` exposes to many methods to add a component to the container. You note that methods `add` of module are confusing, this is worse for `ComponentContainer` imo (look at all the add* methods). In addition, some methods have nothing to do with adding a component and they should not be called from a module.

Comment (simonbrandhof) [inline sonar-core/src/main/java/org/sonar/core/component/Module.java]: If the parameter ComponentContainer is not added to ```configureModule()```, then ```add(Object, boolean)``` and its single usage for MeasureDao should be dropped.

**Comment by sns-seb** [comment]
PR updated:
* rebased on master
* unused protected methods in `Module` removed
* fix `MeasuresDao` not being a singleton

**Comment by simonbrandhof** [comment]
LGTM

**Pull request merged**